Explorer/Assets/Scripts/CRDT/CRDTTests/CRDTSerializerShould.cs

```diff
@@ -0,0 +1,116 @@
+﻿using CRDT.Protocol;
+using CRDT.Protocol.Factory;
+using CRDT.Serializer;
+using NUnit.Framework;
+using System;
+
+namespace CRDT.CRDTTests
+{
+    [TestFixture]
+    public class CRDTSerializerShould
+    {
+        private CRDTSerializer crdtSerializer;
+
+        [SetUp]
+        public void SetUp()
+        {
+            crdtSerializer = new CRDTSerializer();
+        }
+
+        [Test]
+        [TestCase(0, 1, 100, null,
+            ExpectedResult = new byte[]
+            {
+                24, 0, 0, 0,    //msg_length
+                1, 0, 0, 0,     //msg_type
+                0, 0, 0, 0,     //entityId
+                1, 0, 0, 0,     //componentId
+                100, 0, 0, 0,   //timestamp
+                0, 0, 0, 0      //data_length
+            })]
+
+        [TestCase(32424, 67867, 2138996092, new byte[] { 42, 33, 67, 22 },
+            ExpectedResult = new byte[]
+            {
+                28, 0, 0, 0,        //msg_length
+                1, 0, 0, 0,         //msg_type
+                168, 126, 0, 0,     //entityId
+                27, 9, 1, 0,        //componentId
+                124, 125, 126, 127, //timestamp
+                4, 0, 0, 0,         //data_length
+                42, 33, 67, 22      //data
+            })]
+
+        public byte[] SerializeCorrectlyPutComponent(int entityId, int componentId, int timestamp, byte[] data)
+        {
+            ReadOnlyMemory<byte> dataMemory = data ?? ReadOnlyMemory<byte>.Empty;
+            var entity = new CRDTEntity(entityId);
+            var messageDataLength = CRDTMessageSerializationUtils.GetMessageDataLength(CRDTMessageType.PUT_COMPONENT, in dataMemory);
+
+            var message = new CRDTMessage(
+                CRDTMessageType.PUT_COMPONENT,
+                entity,
+                componentId,
+                timestamp,
+                dataMemory
+            );
+
+            var processedMessage = new ProcessedCRDTMessage(message, messageDataLength);
+            var destination = new byte[messageDataLength];
+            var destinationSpan = destination.AsSpan();
+
+            crdtSerializer.Serialize(ref destinationSpan, in processedMessage);
+
+            // No extra bytes allocated
+            Assert.AreEqual(0, destinationSpan.Length);
+
+            return destination;
+        }
+
+        [Test]
+        [TestCase(126, 44, 100,
+            ExpectedResult = new byte[]
+            {
+                20, 0, 0, 0,    //msg_length
```

Comment:
(observation) Do all CRDT messages have this structure? 4 bytes seems wasteful for a type

Reply:
yes, 4 bytes are wasteful for the type, it is the protocol, can't change it unfortunately

Reply:
On the client side I convert it to `byte` via `ReadEnumAs` and `WriteEnumAs`